### ISSUE TITLE
Dataclass Final Member Initialization

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [microsoft/pyright#2906]
**Describe the bug**
Final variables are required to be initialized. If they aren't class variables they are expected to be initialized in constructor. The automatically generated dataclass constructor is not being counted as having assigned that variable.

**To Reproduce**
```python
from dataclasses import dataclass
from typing import Final


class Foo1:
    x: Final[int]

    def __init__(self, x: int) -> None:
        self.x = x


@dataclass
class Foo2:
    x: Final[int]  # Error: "x" is declared as "Final[int]", but value is not assigned.

```

**Expected behavior**
I would expect no errors similar to Foo1.

**VS Code extension or command-line**
pyright 1.1.212

**Comment by GBeauregard** [comment]
Final was not specified with dataclasses in mind and it's an open bug what to do about it:

https://bugs.python.org/issue45384

**Comment by erictraut** [comment]
PEP 591 allows instance variable declarations with a `Final` type to omit the assignment in the case where a value is assigned in the `__inti__` method. For a `dataclass`, the `__init__` method is synthesized, so the assignment is implied. So I agree with @GBeauregard that PEP 591 doesn't discuss how it works with PEP 557, the intended behavior is pretty clear.

I've updated pyright to handle this case correctly. The change will be included in the next release.

For what it's worth, this issue also [exists in mypy](https://github.com/python/mypy/issues/10688).

**Comment by GBeauregard** [comment]
Is it possible that specification of the interaction with Final can be rolled into dataclass transforms documentation, and then the remaining issue is to just allow at runtime for `Final` and `ClassVar` to be used together?

**Comment by erictraut** [comment]
No, this doesn't belong in `dataclass_transform`.

**Comment by GBeauregard** [comment]
Okay, we'll keep it in the `Final` PEP amendments.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.213, which I just published. It will also be included in the next release of pylance.
